src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -359,17 +360,30 @@ void remove_virtual_functionst::get_functions(
   }
 
   // iterate over all children, transitively
-  std::set<irep_idt> visited;
   get_child_functions_rec(
     class_id,
     root_function.symbol_expr,
     component_name,
     functions,
-    visited,
     resolve_function_call);
 
   if(root_function.symbol_expr!=symbol_exprt())
     functions.push_back(root_function);
+
+  // remove all classes from dispatch table where concrete call has been found
+  std::set<irep_idt> concrete_call;
+  for(const auto &entry : functions)
+  {
+    if(entry.symbol_expr != root_function.symbol_expr)
+      concrete_call.insert(entry.class_id);
+  }
+
+  std::remove_if(
+    functions.begin(),
+    functions.end(),
+    [&concrete_call](const dispatch_table_entryt &entry) {
+      return concrete_call.find(entry.class_id) != concrete_call.end();
```

Comment:
Maybe being end-of-the-week dense, but rather than first finding the functions to remove then removing them, can this predicate not simply be `return entry.symbol_expr != root_function.symbol_expr`;?

Reply:
That would also remove those for which the only found `symbol_expr` for the function call is the `root_function.symbol_expr`, wouldn't it? For a class that doesn't implement its own `toString` for example, it makes sense to call `java.lang.Object.toString`. Therefore I collect first entries that do have a more specialized concrete call and remove only those.

But as discussed, it might make sense for example to first have the classes with a more concrete resolved function call and then group those with `root_function.symbol_expr`-"only" into a single conditional to save some GOTO instructions.

For example
```
        // 73 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.StringBuilder" == arg0a->@class_identifier THEN GOTO 3
...
        IF "java::java.util.ArrayList" == arg0a->@class_identifier THEN GOTO 5
        // 58 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.util.Arrays$ArrayList" == arg0a->@class_identifier THEN GOTO 5
...
// all special cases treated, for the rest call `java.lang.Object.toString`
        // 268 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     1: arg0a . java.lang.Object.toString:()Ljava/lang/String;();
        // 269 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.Object.toString:()Ljava/lang/String;#return_value;
...
     3: (struct java.lang.StringBuilder *)arg0a . java.lang.StringBuilder.toString:()Ljava/lang/String;();
        // 277 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.StringBuilder.toString:()Ljava/lang/String;#return_value;
...
        // 284 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     5: (struct java.util.AbstractCollection *)arg0a . java.util.AbstractCollection.toString:()Ljava/lang/String;();
        // 285 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
...
```

Reply:
> That would also remove those for which the only found `symbol_expr` for the function call is the `root_function.symbol_expr`, wouldn't it? For a class that doesn't implement its own `toString` for example, it makes sense to call `java.lang.Object.toString`. Therefore I collect first entries that do have a more specialized concrete call and remove only those.

I still don't understand: `concrete_calls` is created as a set of keys that are a subset of those contained in `functions` (for some loose notion of keys). Then you remove from `functions` all those that have their key in `concrete_calls`, so you remove an element whenever the previous loop inserted it into `concrete_calls`. Thus it seems that
````c++
std::remove_if(
  functions.begin(),
  functions.end(),
  [](const dispatch_table_entryt &entry) {
    entry.symbol_expr != root_function.symbol_expr;
  });
````
would achieve the same, as @thk123 suggested.

What are we missing?